### PR TITLE
listener: add SO_REUSEPORT support for TcpListenSocket by option and UdpListenSocket forcibly

### LINKED ISSUE #8794: Allow REUSEPORT for listener sockets
We should optionally allow listener sockets to be created on each worker with SO_REUSEPORT. This should probably be the default for UDP listeners and opt-in for TCP listeners?

cc @danzh2010 @euroelessar @antoniovicente

### PR DESCRIPTION
**Pull request by l8huang**
This change makes TCP and UDP listeners supporting SO_REUSEPORT.  For TCP,  it can be switched on/off by option 'reuse_port' added to envoy.api.v2.Listener; For UDP, it's always on, the value of 'reuse_port' is ignored.

Signed-off-by: lhuang8 <lhuang8@ebay.com>

Risk Level: Medium
Testing: Update UTs and integration tests, new UT cases added
Docs Changes:
Release Notes:
Fixes #8794

### CODE REVIEW

**Comment by mattklein123** [comment]
@l8huang I think @danzh2010 is already working on this so this might conflict with her work. I will let her chime in.

File: include/envoy/network/listener.h [hunk 1]
```diff
@@ -42,6 +42,14 @@ class ListenerConfig {
   virtual Socket& socket() PURE;
   virtual const Socket& socket() const PURE;
 
+
+  /**
+   * @return bool if 'true', each worker thread creates a socket with SO_REUSEPORT;
+   *         otherwise, 'false'.
+   */
+  virtual bool reusePort() PURE;
+  virtual Network::SocketSharedPtr createReusePortSocket() PURE;
```
[the remark this answers is not in the document]
Comment (l8huang) [reply include/envoy/network/listener.h]: I understood that a socket factory method is needed here for creating reusing port socket, but currently socket() is a simple getter, the socket is created outside of ListenerImpl, if REUSEPORT is not enabled, worker threads need to get the shared socket; if REUSEPORT is enabled and port number is 0, looks like a shared socket is still needed for determining the assigned port number, then worker threads can use that to create reusing port sockets. That's, we need both socket() and a socket factory method. 

No worry, I will resolve the conflicts based on your change after it's ready.

[the remark this answers is not in the document]
Comment (danzh2010) [reply include/envoy/network/listener.h]: What's in my mind is that a not REUSEPORT socket factory provides the shared socket, and a REUSEPORT factory creates individual socket. ListenerConfig only needs to provide access to the socket factory. Socket factory creation is based on config.

**Comment by danzh2010** [comment]
This PR needs to sync with https://github.com/envoyproxy/envoy/pull/8976. After that, do you mind to change this PR to add SO_REUSEPORT support for TcpListenSocket by option and UdpListenSocket all the time?

**Comment by l8huang** [comment]
@danzh2010  sure, I will merge your change and do so ;)

**Comment by mattklein123** [comment]
Can you merge master and then we can take a look? Excited to see this land to get SO_REUSEPORT for UDP listeners!

/wait

**Comment by l8huang** [comment]
@danzh2010 @mattklein123 merged master, could you please take a look?

File: source/server/listener_impl.cc [hunk 2]
```diff
@@ -26,19 +26,36 @@
...
+                                                 const Network::Socket::OptionsSharedPtr& options,
+                                                 bool bind_to_port,
+                                                 const std::string& listener_name, bool reuse_port)
+    : factory_(factory), local_address_(address), socket_type_(socket_type), options_(options),
+      bind_to_port_(bind_to_port), listener_name_(listener_name), reuse_port_(reuse_port) {
+
+  // If reuse_port_ is false, create a socket here which will be used by all worker
+  // threads; otherwise, if port is 0, still need to create a socket here, because
+  // a random port will be assigned, then all worker threads should use same port.
+  if (!reuse_port_ || local_address_->ip()->port() == 0) {
```
Comment (danzh2010) [inline source/server/listener_impl.cc]: nit: ASSERT(socket_type == Stream)

Comment (danzh2010) [inline source/server/listener_impl.cc]: I would like to not create a temporary socket just to get ephemeral port number if possible. Can you try the approach I proposed in my other comment to see if it's possible?

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -567,6 +567,15 @@ void ListenerManagerImpl::addListenerToWorker(Worker& worker, ListenerImpl& list
       }
       if (success) {
         stats_.listener_create_success_.inc();
+
+        if (listener.reusePort()) {
+          // The shared socket created for acquiring port number when port is configured as 0
+          // can be closed here.
```
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: Worker threads run concurrently, main thread posts per-worker listener creation to all workers, the order is not guaranteed.

File: test/server/connection_handler_test.cc [hunk 4]
```diff
@@ -264,7 +267,35 @@ TEST_F(ConnectionHandlerTest, AddDisabledListener) {
   handler_->disableListeners();
   handler_->addListener(*test_listener);
 }
-
+/*
```
[the remark this answers is not in the document]
Comment (l8huang) [reply test/server/connection_handler_test.cc]: this is for my original change, I commented it out and forgot to remove it. Will remove it.

File: source/server/listener_impl.cc [hunk 5]
```diff
@@ -26,19 +26,36 @@
...
+Network::SocketSharedPtr ListenSocketFactoryImpl::createListenSocketAndApplyOptions() {
   // socket might be nullptr depending on factory_ implementation.
-  Network::SocketSharedPtr socket =
-      factory_.createListenSocket(local_address_, socket_type_, options_, bind_to_port_);
+  Network::SocketSharedPtr socket = factory_.createListenSocket(
+      local_address_, socket_type_, options_, bind_to_port_, reuse_port_);
+
   // Binding is done by now.
-  ENVOY_LOG(info, "Create listen socket for listener {} on address {}", listener_name_,
+  ENVOY_LOG(debug, "Create listen socket for listener {} on address {}", listener_name_,
```
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_impl.cc]: this log will output once for each worker thread, if the concurrency is a big value, e.g more than 8, then it becomes too many.

File: include/envoy/network/listener.h [hunk 6]
```diff
@@ -44,8 +44,11 @@ class ListenSocketFactory {
   virtual const Address::InstanceConstSharedPtr& localAddress() const PURE;
 
   /**
-   * @return the socket if getListenSocket() returns a shared socket among each call,
-   * nullopt otherwise.
+   * @return the socket shared by worker threads; if work threads doesn't share a socket,
+   * return null.
+   * Worker threads share socket in two cases:
+   * - listener's 'reuse_port' is set to 'false'
+   * - listener's 'reuse_port' is set to 'true' but port is 0
```
[the remark this answers is not in the document]
Comment (l8huang) [reply include/envoy/network/listener.h]: will update the comment, for the 2nd item, the shared socket is just a ephemeral socket for reserving a port number.

File: source/server/listener_impl.cc (hunk 2 above)
Comment (l8huang) [reply source/server/listener_impl.cc]: socket_type could be Datagram here.

Comment (l8huang) [reply source/server/listener_impl.cc]: on worker thread things happen currently, a real port is reserved with bind() for port 0, I guess there are only two ways to resolve this problem:
- main thread bind port 0 to reserve a port
- worker threads do mutex to guarantee the first one can reserve a port an others use same

We'd better avoid using mutex for reserving a port.

Comment (danzh2010) [reply source/server/listener_impl.cc]: I think Datagram socket will for sure reuse port, no? My point is enforcing this assumption here.

File: source/server/listener_manager_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: Envoy::Server::WorkerImpl::addListener() is called on main thread, so socket can be created in it and posted to worker thread. In this way, Envoy::Network::ConnectionHandler::addListener() should be split into two:

- Envoy::Network::ConnectionHandler::addTcpListener()
- Envoy::Network::ConnectionHandler::addUdpListener()

And ActiveUdpListener and ActiveTcpListener's ctor don't call `onfig.listenSocketFactory().getListenSocket()` to create socket any more.

Sound good?

[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: One point just hit me. Actually even with non-zero port, unless we create all the sockets first and then post to different workers, there is always unstable hashing for connections opened between when the corresponding socket starts listening and when another socket is created. So I think we need to create all the sockets according to concurrency in ListenerManager first before we start listening on any of them if we are reusing port.

File: include/envoy/network/listener.h [hunk 7]
```diff
@@ -83,6 +86,12 @@ class ListenerConfig {
    */
   virtual bool bindToPort() PURE;
 
+  /**
+   * @return bool if 'true', the listener creates one socket with SO_REUSEPORT for each
+   * worker thread; otherwise, all worker threads share one socket.
+   */
+  virtual bool reusePort() PURE;
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply include/envoy/network/listener.h]: > IMO if reusePort is always true for both TCP and UDP, this can be removed, otherwise, it's better to keep this interface. E.g. when calling ListenerComponentFactory::createListenSocket(), reuse_port is needed, if it's true, duplicate socket from hot restart parent process should be skipped.

Socket factory implementation already latches this value and ListenerComponentFactory::createListenSocket() is called from there. So I don't think we need this interface in ListenerConfig. But I think we should finalize the reuse port socket factory implementation first before thinking about what interface it needs to create a socket.

> Actually I think some other methods in Envoy::Server::ListenerImpl can be lift to ListenerConfig interface, e.g address(), socket type, name(). basically theses information are from envoy::api::v2::Listener.

We prefer not to make the interface more complicated unless it's necessary.

File: source/server/listener_impl.cc (hunk 2 above)
Comment (danzh2010) [reply source/server/listener_impl.cc]: Is it cleaner to have two implementations one for providing reuse port socket(UDP or TCP), the other for shared socket(only TCP)? I could foresee that these two implementations will be totally different, so there is no point of mixing them together.

File: source/server/listener_manager_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: For TCP I don't see any issue with the unstable hashing you mentioned, after a connection accepted, a random port  number will be allocated for establishing the connection.
For UDP, packet are stateless datagram, does it matter which socket a packet lands on?

The only issue with ephemeral socket I can see is with UDP, after bind() called(happens during NetworkSocket creation), packets may be delivered to the socket, then after the ephemeral socket closed, the packets looks like dropped on data path, packet being dropped is acceptable for UDP. Even all reuse port UDP sockets got created on main thread and posted to worker threads, before worker threads actually start receiving packets, if the traffic is heavy, some packets may be dropped. Please correct me if I'm wrong.

File: source/server/listener_impl.cc (hunk 2 above)
Comment (l8huang) [reply source/server/listener_impl.cc]: reuse_port is just an option, I don't think it's necessary to have two factory implementations because of it.

Comment (l8huang) [reply source/server/listener_impl.cc]: > nit: ASSERT(socket_type == Stream)

added like:
  if (socket_type_ == Network::Address::SocketType::Datagram) {
    ASSERT(reuse_port_ == true);
  }

File: source/server/listener_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_impl.cc]: The port info can be got by admin API /listeners and /config_dump. In my test env, envoy have 64 worker threads, this log really causes spam.

File: test/server/connection_handler_test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply test/server/connection_handler_test.cc]: removed

File: include/envoy/network/listener.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply include/envoy/network/listener.h]: Simplified interface doc string here, added more details to ListenSocketFactoryImpl:: sharedSocket() doc string.

File: source/server/listener_impl.cc (hunk 2 above)
Comment (l8huang) [reply source/server/listener_impl.cc]: This discussion continues in another comments ;)

File: source/server/listener_manager_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: TCP with a temporary socket might also has some issue. What if a new connection is dispatched to the temporary socket before it's closed? We don't have a listener for that socket.

File: source/server/listener_impl.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_impl.cc]: Okay, sounds fair.

File: source/server/listener_impl.cc (hunk 2 above)
Comment (danzh2010) [reply source/server/listener_impl.cc]: I think we need to figure out how to create socket for reuse port first before looking into these implementation details.

File: source/server/listener_manager_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: > What if a new connection is dispatched to the temporary socket before it's closed? 

the temporary socket doesn't call listen(), no connection will be dispatched to it.

[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: > For UDP, packets are dispatched to one of the reuse_port sockets based on the 4 tuple. The mapping is between the 4 tuple and the index of the socket in a socket map for the reused port. If there are change in that map, the index of each socket might be changed (depending on kernel implementation). So the packet with the same 4 tuple will be mapped to different listening socket.

Even all reuse port UDP sockets are created on main thread at same time, these sockets are bound one by one, during this procedure, the problem you described could happen as well, right? I feel this is becoming a listener readiness issue, that's, before a listener's reuse port sockets are all bound(that's, hash map is stable), no UDP traffic should land on Envoy. 

Looks like QUIC has solution for this issue; for UDP proxying, considering UDP server/client must tolerate packet-dropping, packet-reordering, I guess this doesn't cause real impact in most of cases.

[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: > > What if a new connection is dispatched to the temporary socket before it's closed?
> 
> the temporary socket doesn't call listen(), no connection will be dispatched to it.

TBH, I don't know the kernel behavior for TCP where we have N sockets bound to a port via SO_REUSEPORT and (N - 1) of them are listening. Better to get some kernel person to confirm...

[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: `man listen` has a description about its behavior:

> listen() marks the socket referred to by sockfd as a passive socket,
> that is, as a socket that will be used to accept incoming connection
> requests using accept(2).

see http://man7.org/linux/man-pages/man2/listen.2.html

[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: > Even all reuse port UDP sockets are created on main thread at same time, these sockets are bound one by one, during this procedure, the problem you described could happen as well, right? I feel this is becoming a listener readiness issue, that's, before a listener's reuse port sockets are all bound(that's, hash map is stable), no UDP traffic should land on Envoy.
> Looks like QUIC has solution for this issue; for UDP proxying, considering UDP server/client must tolerate packet-dropping, packet-reordering, I guess this doesn't cause real impact in most of cases.

This might not be a packet loss or reordering issue from proxying Envoy's perspective. Packets from the same client might be proxied by two different worker threads to two different backend services. Because the two listeners on different threads will regard them as two individual traffic flow.

It is a listener readiness issue indeed. If we can guarantee that all the listeners start receiving traffic at the same time after socket binding and cleaning up, yes, opening and closing temporary shared socket in between doesn't have real impact. @mattklein123, @alyssawilk is this guaranteed?

The same problem can happen during listener shutdown too. Because any socket close can cause the socket mapping change. We need to guarantee that sockets are closed at the same time. I feel that UDP SO_REUSEPORT might need special handling at this point.

Current approach seems to work for TCP in my eyes. So if you just want to focus on TCP case, feel free to leave UDP socket as it.

File: source/server/listener_impl.cc [hunk 8]
```diff
@@ -26,19 +26,40 @@
...
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(ListenerComponentFactory& factory,
+                                                 Network::Address::InstanceConstSharedPtr address,
+                                                 Network::Address::SocketType socket_type,
+                                                 const Network::Socket::OptionsSharedPtr& options,
+                                                 bool bind_to_port,
+                                                 const std::string& listener_name, bool reuse_port)
+    : factory_(factory), local_address_(address), socket_type_(socket_type), options_(options),
+      bind_to_port_(bind_to_port), listener_name_(listener_name), reuse_port_(reuse_port) {
+
+  if (socket_type_ == Network::Address::SocketType::Datagram) {
```
Comment (l8huang) [inline source/server/listener_impl.cc]: Should check address type as well, which could be Pipe

File: source/server/listener_impl.cc [hunk 9]
```diff
@@ -26,19 +26,40 @@
...
+    ASSERT(reuse_port_ == true);
+  }
+
+  // If reuse_port_ is false, create a socket here which will be used by all worker
+  // threads; otherwise, if port is 0, still need to create a socket here, because
+  // a random port will be assigned, then all worker threads should use same port.
+  if (!reuse_port_ || local_address_->ip()->port() == 0) {
```
Comment (l8huang) [inline source/server/listener_impl.cc]: If address type is Pipe, shared socket should be always created and reuse_port should be ignored.

File: include/envoy/network/listener.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply include/envoy/network/listener.h]: reusePort() removed from ListenerConfig

File: source/server/listener_manager_impl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: The temporary socket created here will be handed over to first worker thread which calls ListenSocketFactory::getListenSocket() now, that's, no ephemeral port closing anymore.

[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: This sounds safer! Please also document the design somewhere and the rationale behind, including TCP and UDP (if you want to get your hands dirty) initial starts up and tear down, as well as config update. If even it's not supported in current implementation, it's better to mention it at least. I think it's also better to get kernel team looped in for the design.

[the remark this answers is not in the document]
Comment (htuch) [reply source/server/listener_manager_impl.cc]: We need to maintain port 0 ephemeral binding for both integration testing and some use cases, e.g. an Envoy binds port 0, then does service registration with its discovered port.

File: api/envoy/api/v2/lds.proto [hunk 10]
```diff
@@ -238,4 +238,11 @@ message Listener {
   // If no configuration is specified, Envoy will not attempt to balance active connections between
   // worker threads.
   ConnectionBalanceConfig connection_balance_config = 20;
+
+  // When this flag is set to true, listeners sets *SO_REUSEPORT* socket option and
+  // create one socket for each worker thread, this makes inbound connections
+  // distributed among worker threads evenly. When this flag is set to false,
+  // all worker threads shares one socket.
+  // For UDP this flag is set to true forcibly.
+  google.protobuf.BoolValue reuse_port = 21;
```
Comment (htuch) [inline api/envoy/api/v2/lds.proto]: This can probably be just a `bool`, as the default is false (even with the UDP override).

Comment (mattklein123) [reply api/envoy/api/v2/lds.proto]: +1

Here is edited text that I would use:
```
 // When this flag is set to true, listeners set the *SO_REUSEPORT* socket option and
 // create one socket for each worker thread. This makes inbound connections
 // distribute among worker threads roughly evenly in cases where there are a high number
 // of connections from different sources to hash. When this flag is set to false,
 // all worker threads share one socket. For UDP this flag is set to true forcibly.
```

Also, please add a release note with a ref link to this field.

Also, is it still true that when using SO_REUSEPORT for TCP that connections can sometimes be dropped? I know that used to be the case. If so, should we warn about that in hot restart cases?

File: include/envoy/server/listener_manager.h [hunk 11]
```diff
@@ -52,10 +52,9 @@ class ListenerComponentFactory {
    * @param bind_to_port supplies whether to actually bind the socket.
    * @return Network::SocketSharedPtr an initialized and potentially bound socket.
    */
-  virtual Network::SocketSharedPtr
-  createListenSocket(Network::Address::InstanceConstSharedPtr address,
-                     Network::Address::SocketType socket_type,
-                     const Network::Socket::OptionsSharedPtr& options, bool bind_to_port) PURE;
+  virtual Network::SocketSharedPtr createListenSocket(
+      Network::Address::InstanceConstSharedPtr address, Network::Address::SocketType socket_type,
+      const Network::Socket::OptionsSharedPtr& options, bool bind_to_port, bool reuse_port) PURE;
```
Comment (mattklein123) [inline include/envoy/server/listener_manager.h]: Please update the doc above for the new field. Also, I have a strong preference that we replace the 2 bool params with a param struct. This will avoid having to update a ton of different call sites and tests next time we add a param. Since you already have to do it for this change it seems less painful for you to do it now. :)

File: source/server/hot_restarting_parent.cc [hunk 12]
```diff
@@ -100,7 +100,7 @@ HotRestartingParent::Internal::getListenSocketsForChild(const HotRestartMessage:
     Network::ListenSocketFactory& socket_factory = listener.get().listenSocketFactory();
     if (*socket_factory.localAddress() == *addr && listener.get().bindToPort()) {
       if (socket_factory.sharedSocket().has_value()) {
-        // Pass the socket to the new process iff it is already shared across workers.
+        // Pass the socket to the new process if it is already shared across workers.
```
Comment (mattklein123) [inline source/server/hot_restarting_parent.cc]: iff means "if and only if", I would probably revert this change.

File: source/server/http/admin.h [hunk 13]
```diff
@@ -334,19 +334,16 @@ class AdminImpl : public Admin,
     }
 
     Network::SocketSharedPtr getListenSocket() override {
-      // This is only supposed to be called once.
```
Comment (mattklein123) [inline source/server/http/admin.h]: Why is this change needed? The admin server only listens on the main thread.

File: source/server/listener_impl.cc [hunk 14]
```diff
@@ -26,19 +26,53 @@
...
+    if (reuse_port_ == false) {
+      // create a socket which will be used by all worker threads
+      create_socket = true;
+    } else if (local_address_->ip()->port() == 0) {
+      // port is 0, need to create a socket here for reserving a real port number,
+      // then all worker threads should use same port.
+      create_socket = true;
+    }
+  } else {
+    // Listeners with Unix domain socket always use shared socket.
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: Can you ASSERT that type() is Type::Pipe here?

File: source/server/listener_impl.cc [hunk 15]
```diff
@@ -59,42 +93,26 @@ Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApply
   return socket;
 }
 
-void ListenSocketFactoryImplBase::setLocalAddress(
-    Network::Address::InstanceConstSharedPtr local_address) {
-  ENVOY_LOG(debug, "Set listener {} socket factory local address to {}", listener_name_,
-            local_address->asString());
-  local_address_ = local_address;
-}
+Network::SocketSharedPtr ListenSocketFactoryImpl::getListenSocket() {
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: nit: del empty line

File: source/server/listener_impl.cc [hunk 16]
```diff
@@ -59,42 +93,26 @@ Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApply
...
-    : ListenSocketFactoryImplBase(factory, local_address, Network::Address::SocketType::Stream,
-                                  options, bind_to_port, listener_name) {
-  socket_ = createListenSocketAndApplyOptions();
-  if (socket_ != nullptr && localAddress()->ip() != nullptr && localAddress()->ip()->port() == 0) {
-    setLocalAddress(socket_->localAddress());
-  }
-}
+  if (reuse_port_) {
+    Network::SocketSharedPtr socket;
+    absl::call_once(steal_once_, [this, &socket]() {
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: TBH I find this logic to be a little bit confusing. How about we just always store a copy of the socket on the listener config regardless of whether it is reuse port or not, port 0 or not, etc. Then we only need to store a bool of whether it's reuse port or not so the socket should be recreated? As long as we never listen on the socket on the main thread I don't think it will matter if it's sitting there? 

I think this would unify the logic quite a bit. WDYT?

Comment (danzh2010) [reply source/server/listener_impl.cc]: Would it be more clear if we separate these implementations into different factories, I.e. TcpSharedSocketFactory, TcpReusePortSocketFactory, and UdpSocketFactory? TcpSharedSocketFactory is just as the original implementation. TcpReusePortSocketFactory can be implemented with a temporary socket which is transferred to listeners later. UdpSocketFactory should needs more care about socket life time as we want to keep all the sockets open till all listeners stop  listening.

Comment (l8huang) [reply source/server/listener_impl.cc]: IMHO, adding factories for this purpose makes things more complicated.

Comment (mattklein123) [reply source/server/listener_impl.cc]: > For UDP, if a bound socket is not used, kernel still forwards packets to it, but no worker thread handles them, so its socket fd must be closed.

OK, right, that makes sense. Can you add more comments about why we do it this way? It's not clear from the comment trail all of the discussion that was done on this topic and we should make sure it's not lost.

> IMHO, adding factories for this purpose makes things more complicated.

I don't have a strong opinion on this so will defer to @danzh2010. I think we can fix it up later if needed.

File: docs/root/intro/version_history.rst [hunk 17]
```diff
@@ -23,6 +23,7 @@ Version history
 * thrift_proxy: added support for cluster header based routing.
 * tls: remove TLS 1.0 and 1.1 from client defaults
 * router: exposed DOWNSTREAM_REMOTE_ADDRESS as custom HTTP request/response headers.
+* listeners: added :ref:`reuse_port<envoy_api_field_Listener.reuse_port>` option.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: alpha order

File: api/envoy/api/v3alpha/lds.proto [hunk 18]
```diff
@@ -223,4 +223,12 @@ message Listener {
   // If no configuration is specified, Envoy will not attempt to balance active connections between
   // worker threads.
   ConnectionBalanceConfig connection_balance_config = 20;
+
+  // When this flag is set to true, listeners set the *SO_REUSEPORT* socket option and
+  // create one socket for each worker thread. This makes inbound connections
+  // distribute among worker threads roughly evenly in cases where there are a high number
+  // of connections. When this flag is set to false, all worker threads share one socket.
+  // For UDP this flag is set to true forcibly.
+  // Before Linux v4.19-rc1, new TCP connections may be rejected during hot restart.
```
Comment (mattklein123) [inline api/envoy/api/v3alpha/lds.proto]: Might as well ref link to the patch as well as the original issue? Seems useful.

File: source/server/listener_impl.cc [hunk 19]
```diff
@@ -26,19 +26,54 @@
...
-    bool bind_to_port, const std::string& listener_name)
-    : factory_(factory), local_address_(local_address), socket_type_(socket_type),
-      options_(options), bind_to_port_(bind_to_port), listener_name_(listener_name) {}
-
-Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApplyOptions() {
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(ListenerComponentFactory& factory,
+                                                 Network::Address::InstanceConstSharedPtr address,
+                                                 Network::Address::SocketType socket_type,
+                                                 const Network::Socket::OptionsSharedPtr& options,
+                                                 bool bind_to_port,
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: Can you use the same creation param struct here? I might just change `duplicate_parent_socket` to `reuse_port` there for clarity? Up to you.

File: api/envoy/api/v3alpha/lds.proto (hunk 18 above)
Comment (l8huang) [reply api/envoy/api/v3alpha/lds.proto]: done

File: docs/root/intro/version_history.rst (hunk 17 above)
Comment (l8huang) [reply docs/root/intro/version_history.rst]: done

File: source/server/listener_impl.cc [hunk 20]
```diff
@@ -106,6 +129,9 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
       listener_scope_(
           parent_.server_.stats().createScope(fmt::format("listener.{}.", address_->asString()))),
       bind_to_port_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.deprecated_v1(), bind_to_port, true)),
+      socket_type_(Network::Utility::protobufAddressSocketType(config.address())),
+      reuse_port_((socket_type_ == Network::Address::SocketType::Datagram) ? true
```
Comment (danzh2010) [inline source/server/listener_impl.cc]: I highly suspect that we don't need to latch these two parameters. socket_type_ is only used in this constructor and then it is latched in socket factory already. Though reuse_port_ is used in createListenSocketFactory() via ListenerImpl::reusePort(), I think the caller of this method actually has direct access to listener protobuf. Can you try to remove these two members and reusePort() from this impl?

Comment (danzh2010) [reply source/server/listener_impl.cc]: I didn't mean that you always get them from protobuf... These two variables is only needed in this constructor or the caller of it. So they can just be local variable.

File: source/server/listener_impl.h [hunk 21]
```diff
@@ -14,29 +14,44 @@
...
     return local_address_;
   }
 
+  Network::SocketSharedPtr getListenSocket() override;
+
+  /**
+   * @return the socket shared by worker threads; otherwise return null.
+   */
+  absl::optional<std::reference_wrapper<Network::Socket>> sharedSocket() const override {
+    if (!reuse_port_ && socket_) {
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_impl.h]: How can socket_ be null if reuse port? Should it just be ASSERT()?

File: source/server/listener_manager_impl.cc [hunk 22]
```diff
@@ -688,21 +697,17 @@ void ListenerManagerImpl::stopListeners(StopListenersType stop_listeners_type) {
...
-                     }
-                     for (auto& listener : active_listeners_) {
-                       if (listener->listenerTag() == listener_tag) {
-                         listener->listenSocketFactory().sharedSocket()->get().close();
-                       }
-                     }
-                   });
+      stopListener(listener, [this, listener_tag]() {
+        stats_.listener_stopped_.inc();
+        for (auto& listener : active_listeners_) {
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/server/listener_manager_impl.cc]: I would prefer keeping the original behavior if there is no substantial improvement because of this change. If socket is not shared, what the for loop does doesn't make such sense.

File: source/server/listener_impl.cc (hunk 20 above)
Comment (l8huang) [reply source/server/listener_impl.cc]: reasons why they are members:
- ListenerImpl::reusePort() returns `reuse_port_`.
-  `reuse_port_` is defined as const 
- `reuse_port_` must be inited in ctor, it depends on socket_type, so socket_type is lifted as a member
- ListenerManagerImpl::createListenSocketFactory() calls  ListenerImpl::reusePort()  to create ListenSocketFactoryImpl

PS: Just thought if a class should have a member not only determined by in how many methods it's used, socket_type and reuse_port are attributes of a listener, they could be used somewhere later.

File: source/server/listener_impl.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_impl.h]: ASSERT added.

File: source/server/listener_manager_impl.cc (hunk 22 above)
[the remark this answers is not in the document]
Comment (l8huang) [reply source/server/listener_manager_impl.cc]: reverted.

File: source/server/listener_impl.cc [hunk 23]
```diff
@@ -59,42 +94,30 @@ Network::SocketSharedPtr ListenSocketFactoryImplBase::createListenSocketAndApply
...
-  Network::SocketSharedPtr socket = createListenSocketAndApplyOptions();
-  if (socket != nullptr && localAddress()->ip() != nullptr && localAddress()->ip()->port() == 0) {
-    setLocalAddress(socket->localAddress());
+  Network::SocketSharedPtr socket;
+  absl::call_once(steal_once_, [this, &socket]() {
+    if (socket_) {
+      // If a listener's port is set to 0, socket_ should be created for reserving a port
+      // number, it is handed over to the first worker thread came here.
+      // There are several reasons for doing this:
+      // - for UDP, once a socket being bound, it begins to receive packets, it can't be
```
Comment (mattklein123) [inline source/server/listener_impl.cc]: As an aside, I'm not sure how much this really matters. There is going to be a delay between creating the socket on the main thread and when the first worker actually steals it (the delay may be very large due to xDS init, etc.). In general this should be covered by downstream health check, etc. This is related to an offline conversation I had with @danzh2010 about UDP and hot restart. I think it probably will never work correctly with stateless protocols like QUIC. I will open an issue to figure out how to deal with that.

File: source/server/listener_impl.cc (hunk 20 above)
Comment (danzh2010) [reply source/server/listener_impl.cc]: It's a bit confusing to latch same value in two different places. reuse_port_ and socket_type_ is already stored in ListenSocketFactory. And later we might need to maintain them in sync. That's why I want to remove these two members from ListenerImpl. If we need to use them somewhere else, we can access them by listener->listenSocketFactory()->reusePort().

Originally socket_type was a local var. 

The only need of reuse_port here is to add SO_REUSEPORT option. You can just embed `(socket_type_ == Network::Address::SocketType::Datagram) ? true : config.reuse_port()` in the if condition.
You can also change ListenerManagerImpl::createListenSocketFactory() to take reuse_port as parameter, and its caller  ListenerManagerImpl::addOrUpdateListener() should also have access to config. So ListenerImpl doesn't need reusePort() interface.

Comment (mattklein123) [reply source/server/listener_impl.cc]: +1 sounds reasonable to me.

/wait

Comment (l8huang) [reply source/server/listener_impl.cc]: I don't think this causes confusion and need additional effort to sync them. If it really causes any issue, such kind of issue is easy to fix later, let finish this PR, ok? ;)

Comment (mattklein123) [reply source/server/listener_impl.cc]: @l8huang I agree with @danzh2010 on this. Please make the change. Thanks!

/wait

Comment (l8huang) [reply source/server/listener_impl.cc]: @mattklein123 @danzh2010 please take a look, changed according to comment.
